### DISCUSSION TITLE
How/when use getStaticProps

### DISCUSSION THREAD

**Discussion by aralroca** [vercel/next.js#11427] [Help] [answered]
We have an e-commerce with Next.js since the last year. And after seeing the new optimizations of 9.3, we are thinking to migrate from a `getInitialProps` to `getStaticProps`. However, there is something that it's not very clear (at least for me)...

The idea with `getStaticProps` is that the data is fetched at build-time, isn't it? Then, is it needed some cron or some mechanism to re-build when this content change... ?? I looked at the docs but I don't understand how to do that. 

Then, if the content change constantly, so instead of keeping the static approach and doing re-builds, is it better to keep on a server, maybe with `getServerSideProps`? Where is the limit when a resource is considered to serve using `getStaticProps` or with `getServerSideProps`?

**Comment by janus-reith** [comment, marked as answer]
@aralroca I was wondering the same thing, here is how I did it:

I am using dynamic routing and getStaticPaths together with getStaticProps.
The name of my page is something like [slug].js
Now in `getStaticPaths` I have `fallback: true` set, meaning there will be values for this which are not available during buildtime, e.g. /product/mynewproduct.

Next, `getStaticProps` receives that slug as prop from `getStaticPaths` during runtime and fetches my product data for this slug. This only runs once the first time requested, the page served then will be static. You can detect in the render of that page if it currently is in fallback mode, and show a loading spinner or a skeleton. Most of the time you wouldn’t even notice if coming from another page, as that page already triggered a prefetch.

Now in my case, just having the page static the first time it is requested doesn’t really help, as details like price or title might change later.
For that there is a property `revalidate` which you can specify on `getStaticProps` with an amount of seconds, after that amount of seconds the page will be rebuild as new static page.

What is missing for now is a way to programmatically trigger a page revalidate e.g. over an api, so if you have set revalidate set to like 60 seconds, in that 60 seconds you might see outdated product data. For my usecase I simply created a hook which receives the product data from static props and triggers another fetch clientside, and returns the static product data until the client side fetch finished. This is seamless and you have no visible loading time. The approach with the additional client fetch might be unnecessary in your case, but came in handy for me as I wanted to poll clientside for updated availability and prices anyway.

Hope this helped with your considerations!

**Reply by aralroca** [reply]
Thanks for you answer

**Reply by vctormb** [reply]
@janus-reith Interesting technique! Are you using it in prod?
Where did you find this property called `revalidate` in the docs?

**Reply by timneutkens** [reply]
> Where did you find this property called revalidate in the docs?

It's experimental and I've mentioned it once or twice in discussions. I'm currently writing the RFC for it.
Note that as it's experimental the same experimental warnings hold, if you use it you can expect changes in behavior between any version until it's stable.

**Reply by vctormb** [reply]
@timneutkens is there an alternative to update the static page that needs SEO but suffers update constantly?

**Reply by janus-reith** [reply]
@vctormb As @timneutkens already mentioned, I got it from another discussion I created where he wrote about it and linked an example.
I wanted to try it asap, as I was looking for such a feature a long time.
.
I'm using it in production, and it mostly works good so far and every page I checked so far renders and rerenders/reevalidates as expected. The product pages are revalidated every 30 minutes, and in the meantime I'm doing a clientside fetch for product data in a hook which seamlessy replaces the initial data if necessary. However, after requesting several of these pages (~ 1000 pages) by prefetching in a Catalog Grid view, I'm seeing an increase in cpu usage to a varying 20%-60% percent of a core, which doesn't go back to idle anymore.
I currently balance between four nextjs docker containers which share the same static folder as a volume, to make sure each page has to be build only once in the revalidation time frame and not each per container(the cpu usage issue also was there before I did that so thats unrelate). Memory usage starts at ~100mb and goes up to ~600m per container after I had all these pages requested, which seems ok if a lot of pages render, and I had no crash or further issues so far, it`s just a bit uncomfortable to have these seemingly unnecessary server load happening in the background all the time.
I tried so far to make sure all graphql fetch calls happening in getStaticProps have a timeout specified, but that didn't make a difference.

I'm still in the process of figuring out what exactly is happening. If each of the ~1000 pages are revalidated each time after the amount of ms passed I would get that there is constant server load, but @timneutkens as far as I got it, a revalidation is only triggered passively by someone visiting that page after the specfied amount of ms, is that right? Btw, I have a test deployment on zeit now, and while I don't see cpu usage there, maybe this could be interesting for your team aswell to check? (On the other hand, I'd certaily also appreciate any hints regarding debugging if that's not totally out of scope for an experimental feature ;) ) I wonder if the same issues even are present on your serverless approach.

This got a bit long, might create a separate issue for this, but that was my experience so far :)

**Reply by vctormb** [reply]
@janus-reith excellent description about your case! 

Well, we do not have the `revalidate` feature available by now. 
Do you recommend a way to update the data of a generated static page? 
I imagine if I keep the static page with old information and do a fetch in the client side to update the data, the searches engine will get old informations, because I only update the page in the client side, after generating a static page.

**Reply by janus-reith** [reply]
@vctormb Not sure I got the question right, albeit experimental, revalidate is available and in my tests also worked when deployed to Zeit Now.

As you already outlined, just static + Client fetch would result in outdated data, server side generation would have increased response time.
If not using revalidate, you could programmatically  trigger static builds which might be impractical if you have frequent changes, or use server side generation combined with some response cache sitting in front of nextjs.

**Reply by timneutkens** [reply]
RFC posted here: https://github.com/zeit/next.js/discussions/11552

**Reply by saeedPadyab** [reply]
> @aralroca I was wondering the same thing, here is how I did it:
> 
> I am using dynamic routing and getStaticPaths together with getStaticProps.
> The name of my page is something like [slug].js
> Now in `getStaticPaths` I have `fallback: true` set, meaning there will be values for this which are not available during buildtime, e.g. /product/mynewproduct.
> 
> Next, `getStaticProps` receives that slug as prop from `getStaticPaths` during runtime and fetches my product data for this slug. This only runs once the first time requested, the page served then will be static. You can detect in the render of that page if it currently is in fallback mode, and show a loading spinner or a skeleton. Most of the time you wouldn’t even notice if coming from another page, as that page already triggered a prefetch.
> 
> Now in my case, just having the page static the first time it is requested doesn’t really help, as details like price or title might change later.
> For that there is a property `revalidate` which you can specify on `getStaticProps` with an amount of seconds, after that amount of seconds the page will be rebuild as new static page.
> 
> What is missing for now is a way to programmatically trigger a page revalidate e.g. over an api, so if you have set revalidate set to like 60 seconds, in that 60 seconds you might see outdated product data. For my usecase I simply created a hook which receives the product data from static props and triggers another fetch clientside, and returns the static product data until the client side fetch finished. This is seamless and you have no visible loading time. The approach with the additional client fetch might be unnecessary in your case, but came in handy for me as I wanted to poll clientside for updated availability and prices anyway.
> 
> Hope this helped with your considerations!

Hi
Awesome. i am doing same as you said. just about last part, do you use something like useFetch or maybe useSWR?

**Comment by grepug** [comment]
Also I want to know if there is a way to put the generated static HTML to CDN with `fallback: true`.

**Reply by timneutkens** [reply]
If your CDN support it then yes. When using https://zeit.co it's automatically served from the CDN.

**Comment by hardfire** [comment]
I have been playing around with next.js and am using WordPress as a data backend. Using the WordPress rest API I'm building static html pages using `getStaticProps` , `getStaticPaths` and `next export`.
Since the data in `getStaticProps` is built and rendered via node.js on build time, can we disable the JSON props in `<script id="__NEXT_DATA__" type="application/json">` so that the page is slimmer as i don't think in our case the json would be used.
